src/libxtp/aobasis.cc

```diff
@@ -65,10 +102,158 @@ void AOBasis::Fill(const BasisSet& bs, const QMMolecule& atoms) {
       aoshell.CalcMinDecay();
       aoshell.normalizeContraction();
     }
-    _FuncperAtom.push_back(atomfunc);
   }
+  FillFuncperAtom();
   return;
 }
 
+void AOBasis::FillFuncperAtom() {
+  _FuncperAtom.clear();
+  Index currentIndex = -1;
+  for (const auto& shell : _aoshells) {
+    if (shell.getAtomIndex() == currentIndex) {
+      _FuncperAtom[shell.getAtomIndex()] += shell.getNumFunc();
+    } else {
+      currentIndex = shell.getAtomIndex();
+      _FuncperAtom.push_back(shell.getNumFunc());
+    }
+  }
+}
+std::vector<libint2::Shell> AOBasis::GenerateLibintBasis() const {
+  std::vector<libint2::Shell> libintshells;
+  libintshells.reserve(_aoshells.size());
+  for (const auto& shell : _aoshells) {
+    libintshells.push_back(shell.LibintShell());
+  }
+  return libintshells;
+}
+
+std::vector<std::vector<Index>> AOBasis::ComputeShellPairs(
+    double threshold) const {
+
+  Index nthreads = OPENMP::getMaxThreads();
+
+  std::vector<libint2::Shell> shells = GenerateLibintBasis();
+
+  // construct the 2-electron repulsion integrals engine
+  std::vector<libint2::Engine> engines;
+  engines.reserve(nthreads);
+  engines.emplace_back(libint2::Operator::overlap, getMaxNprim(), getMaxL(), 0);
+  for (Index i = 1; i != nthreads; ++i) {
+    engines.push_back(engines[0]);
+  }
+
+  std::vector<std::vector<Index>> pairs(shells.size());
+
+#pragma omp parallel for schedule(dynamic)
+  for (Index s1 = 0; s1 < Index(shells.size()); ++s1) {
+    Index thread_id = OPENMP::getThreadId();
+
+    libint2::Engine& engine = engines[thread_id];
+    const libint2::Engine::target_ptr_vec& buf = engine.results();
+    Index n1 = shells[s1].size();
+
+    for (Index s2 = 0; s2 <= s1; ++s2) {
```

Comment:
I think this is the performance bottleneck, we process the small parts first and the large parts at the end. @rubengerritsen

Reply:
we should let s2=s1; s2<=shells.size

Reply:
but that messes with the ERIs